plutus-core/plutus-core/src/PlutusCore/Builtin/KnownTypeAst.hs

```diff
@@ -229,29 +229,29 @@ toTyNameAst _ =
 instance uni `Contains` f => KnownTypeAst uni (BuiltinHead f) where
     type IsBuiltin (BuiltinHead f) = 'True
     type ToHoles (BuiltinHead f) = '[]
-    type ToBinds (BuiltinHead f) = '[]
+    type ToBinds acc (BuiltinHead f) = acc
     toTypeAst _ = mkTyBuiltin @_ @f ()
     {-# INLINE toTypeAst #-}
 
 instance (KnownTypeAst uni a, KnownTypeAst uni b) => KnownTypeAst uni (a -> b) where
     type IsBuiltin (a -> b) = 'False
     type ToHoles (a -> b) = '[TypeHole a, TypeHole b]
-    type ToBinds (a -> b) = Merge (ToBinds a) (ToBinds b)
+    type ToBinds acc (a -> b) = ToBinds (ToBinds acc a) b
     toTypeAst _ = TyFun () (toTypeAst $ Proxy @a) (toTypeAst $ Proxy @b)
     {-# INLINE toTypeAst #-}
 
 instance (name ~ 'TyNameRep text uniq, KnownSymbol text, KnownNat uniq) =>
             KnownTypeAst uni (TyVarRep name) where
     type IsBuiltin (TyVarRep name) = 'False
     type ToHoles (TyVarRep name) = '[]
-    type ToBinds (TyVarRep name) = '[ 'GADT.Some name ]
+    type ToBinds acc (TyVarRep name) = Insert ('GADT.Some name) acc
     toTypeAst _ = TyVar () . toTyNameAst $ Proxy @('TyNameRep text uniq)
     {-# INLINE toTypeAst #-}
 
 instance (KnownTypeAst uni fun, KnownTypeAst uni arg) => KnownTypeAst uni (TyAppRep fun arg) where
     type IsBuiltin (TyAppRep fun arg) = IsBuiltin fun && IsBuiltin arg
     type ToHoles (TyAppRep fun arg) = '[RepHole fun, RepHole arg]
-    type ToBinds (TyAppRep fun arg) = Merge (ToBinds fun) (ToBinds arg)
+    type ToBinds acc (TyAppRep fun arg) = ToBinds (ToBinds acc fun) arg
```

Comment:
Merging lists each time we stumble upon a type application or a `->` was quite inefficient. Now we just pass an accumulator around and only touch it when there's a variable or a binder, which happens fairly rarely (the latter never happens for the default builtins).